pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -916,33 +995,43 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 	stat.ProcessingTime.End(true)
 	stat.UpstreamTime.Start()
 
+	var ipFamily ipfamily.IPFamily
+	if targetServerAddr.Is4() {
+		ipFamily = ipfamily.IPv4()
+	} else {
+		ipFamily = ipfamily.IPv6()
+	}
+
 	dialer := net.Dialer{
 		Timeout: 2 * time.Second,
 		Control: func(network, address string, c syscall.RawConn) error {
 			var soerr error
 			if err := c.Control(func(su uintptr) {
-				soerr = setSoMark(int(su), ep.GetIdentity())
+				soerr = setSoMarks(int(su), ipFamily, ep.GetIdentity())
 			}); err != nil {
 				return err
 			}
 			return soerr
 		},
 	}
-	client.Dialer = &dialer
 
-	conn, err := client.Dial(targetServerAddrStr)
-	if err != nil {
-		err := fmt.Errorf("failed to dial connection to %v: %w", targetServerAddrStr, err)
-		stat.Err = err
-		scopedLog.WithError(err).Error("Failed to dial connection to the upstream DNS server, cannot service DNS request")
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerID, targetServerAddrStr, request, protocol, false, &stat)
-		p.sendRefused(scopedLog, w, request)
-		return
+	var key string
+	if !ep.IsHost() && !epAddr.IsLoopback() && !targetServerIsWorld && !targetServerID.HasLocalScope() && !targetServerIsLocal {
```

Comment:
The `targetServerID.HasLocalScope()` seems to test if the target is outside of the cluster by checking if it is a CIDR identity. However, these days, CIDR identites can [already overlap with node identities](https://github.com/cilium/cilium/pull/27464) and soon will probably also overlap with pod identities. So `targetServerID.HasLocalScope()` might be true for remote pods in the future. I think we ought to introduce a proper helper here that is a bit more future proof.

Do we have CI coverage for those cases (i.e. testing DNS traffic to a pod, to a CIDR identity, to a remote node)? Otherwise I'm afraid we might introduce regressions, because "local identity" might no longer mean "outside of cluster".

Reply:
We likely do have CI coverage gaps here. Added a new helper `identity.IsCluster()` to keep this semantic in the identity package.